File: src/config-file.ts
import * as fs from 'fs'
import * as path from 'path'
import * as core from '@actions/core'
import * as yaml from 'js-yaml'
import * as utils from './utils.js'

/**
 * Configuration interface for engagement scoring weights
 */
export interface ConfigFileEngagementWeights {
  comments: number
  reactions: number
  contributors: number
  lastActivity: number
  issueAge: number
  linkedPullRequests: number
}

/**
 * Configuration interface for batch label processing
 */
export interface ConfigFileEngagement {
  weights: ConfigFileEngagementWeights
}

/**
 * Configuration interface for a single label group
 */
export interface ConfigFileLabelGroup {
  /** Label prefix for searching (e.g., 'area-', 'platform-') */
  labelPrefix?: string
  /** Template to use for this label group */
  template: string
  /** Specific label to use (for regression type) */
  label?: string
  /** Multiple specific labels to choose from (e.g., for missing-info type) */
  labels?: string[]
}

/**
 * Configuration interface for batch label processing
 */
export interface ConfigFileLabels {
  groups: Record<string, ConfigFileLabelGroup>
}

/**
 * Configuration interface for the .triagerc.yml file
 */
export interface ConfigFile {
  engagement: ConfigFileEngagement
  labels: ConfigFileLabels
}

/**
 * Default engagement weights that match the current hardcoded values
 */
export const DEFAULT_ENGAGEMENT_WEIGHTS: ConfigFileEngagementWeights = {
  comments: 3,
  reactions: 1,
  contributors: 2,
  lastActivity: 1,
  issueAge: 1,
  linkedPullRequests: 2
}

/**
 * Load full triage configuration from .triagerc.yml or .github/.triagerc.yml
 * @param workspacePath - The workspace path to search for config files
 * @returns The full configuration object
 */
export async function loadConfigFile(workspacePath: string = '.'): Promise<ConfigFile> {
  // Validate and normalize workspace path to prevent directory traversal
  const currentDir = process.cwd()
  const normalizedWorkspace = path.resolve(workspacePath)

  // Ensure workspace path is within or at the current working directory
  const relativeToCurrentDir = path.relative(currentDir, normalizedWorkspace)
  if (relativeToCurrentDir.startsWith('..') || path.isAbsolute(relativeToCurrentDir)) {
    throw new Error(`Invalid workspace path: ${workspacePath} resolves outside current directory`)
  }

  const configPaths = [
    utils.safePath(normalizedWorkspace, '.triagerc.yml'),
    utils.safePath(normalizedWorkspace, '.github/.triagerc.yml')
  ]

  const config = await loadFile(configPaths)

  // Log final config
  core.info(`Using complete configuration: ${JSON.stringify(config)}`)

  return config
}

export async function loadFile(configPaths: string[]): Promise<ConfigFile> {
  const failedPaths: Map<string, string> = new Map()

  for (const configPath of configPaths) {
    try {
      core.info(`Attempting to load triage configuration from ${configPath}`)
      const fileContent = await fs.promises.readFile(configPath, 'utf8')
      const parsedConfig = parseConfigFile(fileContent)
      if (parsedConfig) {
        core.info(`Successfully loaded configuration from ${configPath}: ${JSON.stringify(parsedConfig)}`)
        return parsedConfig
      }
    } catch (error) {
      failedPaths.set(configPath, error instanceof Error ? error.message : 'Unknown error')
    }
  }

  // Log failed lookup
  if (failedPaths.size > 0) {
    const details = Array.from(failedPaths.entries())
      .map(([configPath, errorMessage]) => ` - ${configPath}: ${errorMessage}`)
      .join('\n')
    core.warning(`Failed to load configuration from the following paths:\n${details}`)
  }

  // Nothing was loaded, so we got an empty config file
  return {
    engagement: {
      weights: { ...DEFAULT_ENGAGEMENT_WEIGHTS }
    },
    labels: {
      groups: {}
    }
  }
}

export function parseConfigFile(fileContent: string): ConfigFile | undefined {
  const parsedConfig = yaml.load(fileContent) as Partial<ConfigFile> | undefined

  // We successfully loaded a configuration
  if (!parsedConfig || typeof parsedConfig !== 'object') {
    return undefined
  }

  // Make sure the config is normalized
  const normalized: ConfigFile = {
    ...parsedConfig,
    engagement: {
      ...parsedConfig.engagement,
      weights: {
        ...DEFAULT_ENGAGEMENT_WEIGHTS,
        ...(parsedConfig.engagement?.weights ?? {})
      }
    },
    labels: {
      ...parsedConfig.labels,
      groups: parsedConfig.labels?.groups ?? {}
    }
  }

  return normalized
}

File: src/TriageAssistant.Core/Configuration/ConfigFile.cs
using System.Text.Json.Serialization;
using YamlDotNet.Serialization;

namespace TriageAssistant.Core.Configuration;

/// <summary>
/// Configuration interface for engagement scoring weights
/// </summary>
public class ConfigFileEngagementWeights
{
    [YamlMember(Alias = "comments")]
    [JsonPropertyName("comments")]
    public double Comments { get; set; } = 3;

    [YamlMember(Alias = "reactions")]
    [JsonPropertyName("reactions")]
    public double Reactions { get; set; } = 1;

    [YamlMember(Alias = "contributors")]
    [JsonPropertyName("contributors")]
    public double Contributors { get; set; } = 4;

    [YamlMember(Alias = "lastActivity")]
    [JsonPropertyName("lastActivity")]
    public double LastActivity { get; set; } = 1;

    [YamlMember(Alias = "issueAge")]
    [JsonPropertyName("issueAge")]
    public double IssueAge { get; set; } = 1;

    [YamlMember(Alias = "linkedPullRequests")]
    [JsonPropertyName("linkedPullRequests")]
    public double LinkedPullRequests { get; set; } = 2;
}

/// <summary>
/// Configuration interface for engagement scoring
/// </summary>
public class ConfigFileEngagement
{
    [YamlMember(Alias = "weights")]
    [JsonPropertyName("weights")]
    public ConfigFileEngagementWeights Weights { get; set; } = new();
}

/// <summary>
/// Configuration interface for a single label group
/// </summary>
public class ConfigFileLabelGroup
{
    /// <summary>
    /// Label prefix for searching (e.g., 'area-', 'platform-')
    /// </summary>
    [YamlMember(Alias = "labelPrefix")]
    [JsonPropertyName("labelPrefix")]
    public string? LabelPrefix { get; set; }

    /// <summary>
    /// Template to use for this label group
    /// </summary>
    [YamlMember(Alias = "template")]
    [JsonPropertyName("template")]
    public string Template { get; set; } = string.Empty;

    /// <summary>
    /// Specific label to use (for regression type)
    /// </summary>
    [YamlMember(Alias = "label")]
    [JsonPropertyName("label")]
    public string? Label { get; set; }
}

/// <summary>
/// Configuration interface for batch label processing
/// </summary>
public class ConfigFileLabels
{
    [YamlMember(Alias = "groups")]
    [JsonPropertyName("groups")]
    public Dictionary<string, ConfigFileLabelGroup> Groups { get; set; } = new();
}

/// <summary>
/// Configuration interface for the .triagerc.yml file
/// </summary>
public class ConfigFile
{
    [YamlMember(Alias = "engagement")]
    [JsonPropertyName("engagement")]
    public ConfigFileEngagement Engagement { get; set; } = new();

    [YamlMember(Alias = "labels")]
    [JsonPropertyName("labels")]
    public ConfigFileLabels Labels { get; set; } = new();
}